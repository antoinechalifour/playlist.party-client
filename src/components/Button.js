import styled, { css } from 'styled-components'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'

const typeThemes = {
  primary: css`
    background: ${({ theme }) => theme.colors.primary};
    color: ${({ theme }) => theme.colors.typography};
    box-shadow: 0 1px 6px rgba(0, 0, 0, .45);
    text-decoration: none;
  `,
  secondary: css`
    color: ${({ theme }) => theme.colors.typography};
    box-shadow: 0 1px 6px rgba(0, 0, 0, .35);
    text-decoration: none;
  `,
  tertiary: css`
    color: ${({ theme }) => theme.colors.typography};
    text-decoration: underline;
  `
}

export const Button = styled.button`
  padding: 12px 16px;
  cursor: pointer;
  font-size: inherit;
  background: none;
  border-radius: 4px;

  ${({ variant }) => typeThemes[variant]}
`

Button.propTypes = {
  variant: PropTypes.oneOf(['primary', 'secondary', 'tertiary'])
}

Button.defaultProps = {
  variant: 'tertiary'
}

export const LinkButton = Button.withComponent(Link)

export const ExternalLinkButton = Button.withComponent('a')
