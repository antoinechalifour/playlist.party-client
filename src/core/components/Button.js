import styled, { css } from 'styled-components'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'

const typeThemes = {
  primary: css`
    background: ${({ theme }) => theme.colors.primary};
    border: 1px solid rgba(0, 0, 0, .15);
    color: #fff;
    text-decoration: none;
  `,
  secondary: css`
    color: ${({ theme }) => theme.colors.typography};
    border: 1px solid rgba(0, 0, 0, .15);
    text-decoration: none;
  `,
  tertiary: css`
    color: ${({ theme }) => theme.colors.typography};
    text-decoration: underline;
  `
}

export const Button = styled.button`
  display: inline-block;
  font-size: inherit;
  border: none;
  background: none;
  border-radius: 4px;
  text-align: center;
  padding: 12px 16px;
  cursor: pointer;
  min-width: 150px;

  ${({ variant }) => typeThemes[variant]}
`

Button.propTypes = {
  variant: PropTypes.oneOf(['primary', 'secondary', 'tertiary'])
}

Button.defaultProps = {
  variant: 'tertiary'
}

export const LinkButton = Button.withComponent(Link)

export const ExternalLinkButton = Button.withComponent('a')
