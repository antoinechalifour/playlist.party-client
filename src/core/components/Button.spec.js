import React from 'react'
import renderer from 'react-test-renderer'
import { BrowserRouter } from 'react-router-dom'
import { Button, LinkButton, ExternalLinkButton } from './Button'
import 'jest-styled-components'

const theme = {
  colors: {
    dark: '#151515',
    danger: '#a63d40',
    primary: '#1db954',
    confirm: '#90a959',
    accent: '#6494aa',
    typography: '#373d3f',
    typographyLight: '#bcbcbc'
  }
}

describe('Button', () => {
  it('Should render a primary button', () => {
    const tree = renderer
      .create(<Button theme={theme} variant='primary'>Click me!</Button>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Should render a secondary button', () => {
    const tree = renderer
      .create(<Button theme={theme} variant='secondary'>Click me!</Button>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Should render a tertiary button', () => {
    const tree = renderer
      .create(<Button theme={theme} variant='tertiary'>Click me!</Button>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Should render a tertiary button by default', () => {
    const tree = renderer
      .create(<Button theme={theme}>Click me!</Button>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('LinkButton', () => {
  it('Should render a primary button', () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <LinkButton theme={theme} variant='primary' to='/'>
            Click me!
          </LinkButton>
        </BrowserRouter>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Should render a secondary button', () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <LinkButton theme={theme} variant='secondary' to='/'>
            Click me!
          </LinkButton>
        </BrowserRouter>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Should render a tertiary button', () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <LinkButton theme={theme} variant='tertiary' to='/'>
            Click me!
          </LinkButton>
        </BrowserRouter>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Should render a tertiary button by default', () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <LinkButton theme={theme} to='/'>Click me!</LinkButton>
        </BrowserRouter>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('ExternalLinkButton', () => {
  it('Should render a primary button', () => {
    const tree = renderer
      .create(
        <ExternalLinkButton href='/test' theme={theme} variant='primary'>
          Click me!
        </ExternalLinkButton>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Should render a secondary button', () => {
    const tree = renderer
      .create(
        <ExternalLinkButton href='/test' theme={theme} variant='secondary'>
          Click me!
        </ExternalLinkButton>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Should render a tertiary button', () => {
    const tree = renderer
      .create(
        <ExternalLinkButton href='/test' theme={theme} variant='tertiary'>
          Click me!
        </ExternalLinkButton>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Should render a tertiary button by default', () => {
    const tree = renderer
      .create(
        <ExternalLinkButton href='/test' theme={theme}>
          Click me!
        </ExternalLinkButton>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
