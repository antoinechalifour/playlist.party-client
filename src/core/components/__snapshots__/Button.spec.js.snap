// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Should render a primary button 1`] = `
.c0 {
  display: inline-block;
  font-size: inherit;
  border: none;
  background: none;
  border-radius: 4px;
  text-align: center;
  padding: 12px 16px;
  cursor: pointer;
  min-width: 150px;
  background: #1db954;
  border: 1px solid rgba(0,0,0,.15);
  color: #fff;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<button
  className="c0"
>
  Click me!
</button>
`;

exports[`Button Should render a secondary button 1`] = `
.c0 {
  display: inline-block;
  font-size: inherit;
  border: none;
  background: none;
  border-radius: 4px;
  text-align: center;
  padding: 12px 16px;
  cursor: pointer;
  min-width: 150px;
  color: #373d3f;
  border: 1px solid rgba(0,0,0,.15);
  -webkit-text-decoration: none;
  text-decoration: none;
}

<button
  className="c0"
>
  Click me!
</button>
`;

exports[`Button Should render a tertiary button 1`] = `
.c0 {
  display: inline-block;
  font-size: inherit;
  border: none;
  background: none;
  border-radius: 4px;
  text-align: center;
  padding: 12px 16px;
  cursor: pointer;
  min-width: 150px;
  color: #373d3f;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<button
  className="c0"
>
  Click me!
</button>
`;

exports[`Button Should render a tertiary button by default 1`] = `
.c0 {
  display: inline-block;
  font-size: inherit;
  border: none;
  background: none;
  border-radius: 4px;
  text-align: center;
  padding: 12px 16px;
  cursor: pointer;
  min-width: 150px;
  color: #373d3f;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<button
  className="c0"
>
  Click me!
</button>
`;

exports[`ExternalLinkButton Should render a primary button 1`] = `
.c0 {
  display: inline-block;
  font-size: inherit;
  border: none;
  background: none;
  border-radius: 4px;
  text-align: center;
  padding: 12px 16px;
  cursor: pointer;
  min-width: 150px;
  background: #1db954;
  border: 1px solid rgba(0,0,0,.15);
  color: #fff;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<a
  className="c0"
  href="/test"
>
  Click me!
</a>
`;

exports[`ExternalLinkButton Should render a secondary button 1`] = `
.c0 {
  display: inline-block;
  font-size: inherit;
  border: none;
  background: none;
  border-radius: 4px;
  text-align: center;
  padding: 12px 16px;
  cursor: pointer;
  min-width: 150px;
  color: #373d3f;
  border: 1px solid rgba(0,0,0,.15);
  -webkit-text-decoration: none;
  text-decoration: none;
}

<a
  className="c0"
  href="/test"
>
  Click me!
</a>
`;

exports[`ExternalLinkButton Should render a tertiary button 1`] = `
.c0 {
  display: inline-block;
  font-size: inherit;
  border: none;
  background: none;
  border-radius: 4px;
  text-align: center;
  padding: 12px 16px;
  cursor: pointer;
  min-width: 150px;
  color: #373d3f;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<a
  className="c0"
  href="/test"
>
  Click me!
</a>
`;

exports[`ExternalLinkButton Should render a tertiary button by default 1`] = `
.c0 {
  display: inline-block;
  font-size: inherit;
  border: none;
  background: none;
  border-radius: 4px;
  text-align: center;
  padding: 12px 16px;
  cursor: pointer;
  min-width: 150px;
  color: #373d3f;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<a
  className="c0"
  href="/test"
>
  Click me!
</a>
`;

exports[`LinkButton Should render a primary button 1`] = `
.c0 {
  display: inline-block;
  font-size: inherit;
  border: none;
  background: none;
  border-radius: 4px;
  text-align: center;
  padding: 12px 16px;
  cursor: pointer;
  min-width: 150px;
  background: #1db954;
  border: 1px solid rgba(0,0,0,.15);
  color: #fff;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<a
  className="c0"
  href="/"
  onClick={[Function]}
  theme={
    Object {
      "colors": Object {
        "accent": "#6494aa",
        "confirm": "#90a959",
        "danger": "#a63d40",
        "dark": "#151515",
        "primary": "#1db954",
        "typography": "#373d3f",
        "typographyLight": "#bcbcbc",
      },
    }
  }
  variant="primary"
>
  Click me!
</a>
`;

exports[`LinkButton Should render a secondary button 1`] = `
.c0 {
  display: inline-block;
  font-size: inherit;
  border: none;
  background: none;
  border-radius: 4px;
  text-align: center;
  padding: 12px 16px;
  cursor: pointer;
  min-width: 150px;
  color: #373d3f;
  border: 1px solid rgba(0,0,0,.15);
  -webkit-text-decoration: none;
  text-decoration: none;
}

<a
  className="c0"
  href="/"
  onClick={[Function]}
  theme={
    Object {
      "colors": Object {
        "accent": "#6494aa",
        "confirm": "#90a959",
        "danger": "#a63d40",
        "dark": "#151515",
        "primary": "#1db954",
        "typography": "#373d3f",
        "typographyLight": "#bcbcbc",
      },
    }
  }
  variant="secondary"
>
  Click me!
</a>
`;

exports[`LinkButton Should render a tertiary button 1`] = `
.c0 {
  display: inline-block;
  font-size: inherit;
  border: none;
  background: none;
  border-radius: 4px;
  text-align: center;
  padding: 12px 16px;
  cursor: pointer;
  min-width: 150px;
  color: #373d3f;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<a
  className="c0"
  href="/"
  onClick={[Function]}
  theme={
    Object {
      "colors": Object {
        "accent": "#6494aa",
        "confirm": "#90a959",
        "danger": "#a63d40",
        "dark": "#151515",
        "primary": "#1db954",
        "typography": "#373d3f",
        "typographyLight": "#bcbcbc",
      },
    }
  }
  variant="tertiary"
>
  Click me!
</a>
`;

exports[`LinkButton Should render a tertiary button by default 1`] = `
.c0 {
  display: inline-block;
  font-size: inherit;
  border: none;
  background: none;
  border-radius: 4px;
  text-align: center;
  padding: 12px 16px;
  cursor: pointer;
  min-width: 150px;
  color: #373d3f;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<a
  className="c0"
  href="/"
  onClick={[Function]}
  theme={
    Object {
      "colors": Object {
        "accent": "#6494aa",
        "confirm": "#90a959",
        "danger": "#a63d40",
        "dark": "#151515",
        "primary": "#1db954",
        "typography": "#373d3f",
        "typographyLight": "#bcbcbc",
      },
    }
  }
  variant="tertiary"
>
  Click me!
</a>
`;
